### PR TITLE
use session specific locks instead of global ones

### PR DESCRIPTION
**Pull request by czue**
can be read commit by commit. 

the main change is actually pretty simple/straighforward: https://github.com/dimagi/touchforms/commit/4229b4c66937c98f82a7a6976607bc4d91958cd7. rest is cleanup and removal of code that was either not used at all or only used by offline mode.

@snopoke @benrudolph @TylerSheffels 

it could be good to give this a quick QA pass, though I was pretty careful, ran every line of code that I changed as well as validate the session-specific locks worked as expected. cc @sheelio

### CODE REVIEW

File: touchforms/backend/xformplayer.py [hunk 1]
```diff
@@ -741,14 +726,10 @@ def prev_event (xfsess):
         at_start, ev = True, xfsess.next_event()
     return at_start, ev
 
-def save_form (xfsess, persist=False):
+def save_form(xfsess):
```
Comment (dimagimon) [inline touchforms/backend/xformplayer.py]: E302 expected 2 blank lines, found 1

**Comment by benrudolph** [comment]
:bear: <-- closest thing to beast in the emoticons
this change makes a lot of sense to me. i'm happy to merge when you feel like it's g2g

File: touchforms/backend/xformplayer.py [hunk 2]
```diff
@@ -83,42 +85,24 @@ def get_session(self, session_id, override_state=None):
...
             for sess_id, sess in self.session_cache.items():
                 if now - sess.last_activity > sess.staleness_window:
                     del self.session_cache[sess_id]
                     num_sess_purged += 1
+                    # also purge the session lock
+                    if sess_id in self.session_locks:
+                        with self.session_locks[sess_id]:
+                            del self.session_locks[sess_id]
                 else:
                     num_sess_active += 1
```
Comment (snopoke) [inline touchforms/backend/xformplayer.py]: Don't need this anymore

**Comment by snopoke** [comment]
:tada:

File: touchforms/backend/xformplayer.py [hunk 3]
```diff
@@ -83,42 +85,24 @@ def get_session(self, session_id, override_state=None):
...
     def purge(self):
         num_sess_purged = 0
         num_sess_active = 0
 
-        with self.lock:
+        with self.global_lock:
             now = time.time()
             for sess_id, sess in self.session_cache.items():
                 if now - sess.last_activity > sess.staleness_window:
                     del self.session_cache[sess_id]
```
Comment (czue) [inline touchforms/backend/xformplayer.py]: good call, updated.

**Comment by czue** [comment]
@benrudolph fyi, maybe keep an eye on touchforms when you wake up in case something goes horribly wrong (hopefully unlikely)

**Comment by benrudolph** [comment]
:+1:

**Pull request merged**